nix/test/rebuild/rebuild.nix

```diff
@@ -0,0 +1,331 @@
+        )
+        node2.succeed(
+            "mayastor-client -a ${node2ip} replica create pool2 ${replica2_uuid} --size 100MiB --protocol nvmf"
+        )
+
+        # Create nexus on node 1
+        node1.succeed(
+            "mayastor-client -a ${node1ip} nexus create ${nexus_uuid} 100MiB '{}'".format(
+                get_replica1_uri()
+            )
+        )
+        # Add a second child to nexus and don't start rebuilding
+        node1.succeed(
+            "mayastor-client -a ${node1ip} nexus add ${nexus_uuid} '{}' true".format(
+                get_replica2_uri()
+            )
+        )
+
+
+    def teardown():
+        # Destroy nexus
+        node1.succeed(
+            "mayastor-client -a ${node1ip} nexus destroy ${nexus_uuid}"
+        )
+        # Delete replicas
+        node1.succeed(
+            "mayastor-client -a ${node1ip} replica destroy ${replica1_uuid}"
+        )
+        node2.succeed(
+            "mayastor-client -a ${node2ip} replica destroy ${replica2_uuid}"
+        )
+
+        node1.shutdown()
+        node2.shutdown()
+
+
+    def new_env():
+        teardown()
+        startup()
+
+
+    """
+    Test cases
+    """
+
+    startup()
+    replica2_uri = get_replica2_uri()
+
+    with subtest("start a rebuild"):
+        exec_rebuild_operation("start", replica2_uri)
+        rebuild_state = get_rebuild_state(replica2_uri)
+        expected_state = "running"
+        assert rebuild_state == expected_state, "Rebuild state {}, expected {}".format(
+            rebuild_state, expected_state
+        )
+
+    with subtest("pause a rebuild"):
+        exec_rebuild_operation("pause", replica2_uri)
+
+        # Wait for the rebuild to pause.
+        # We retry because this may take a little time (the rebuild must complete outstanding I/Os before becoming paused).
+        retries = 5
+        while get_rebuild_state(replica2_uri) != "paused" and retries > 0:
+            sleep(1)
+            retries -= 1
+
+        rebuild_state = get_rebuild_state(replica2_uri)
+        expected_state = "paused"
+        assert rebuild_state == expected_state, "Rebuild state {}, expected {}".format(
+            rebuild_state, expected_state
+        )
+
+    with subtest("resume a rebuild"):
+        exec_rebuild_operation("resume", replica2_uri)
+        rebuild_state = get_rebuild_state(replica2_uri)
+        expected_state = "running"
+        assert rebuild_state == expected_state, "Rebuild state {}, expected {}".format(
+            rebuild_state, expected_state
+        )
+
+    with subtest("get number rebuilds"):
+        num_rebuilds = get_num_rebuilds()
+        expected_num_rebuilds = 1
+        assert (
+            num_rebuilds == expected_num_rebuilds
+        ), "Number of rebuilds {}, expected {}".format(num_rebuilds, expected_num_rebuilds)
+
+    with subtest("network fault"):
+        # Wait for some rebuild progress
+        while get_rebuild_progress_percentage() < 5:
+            sleep(1)
```

Comment:
Should this sleep be smaller, eg is it possible that the rebuild completes within this 1 second?

Reply:
The volume has been chosen to be 100MiB so that the rebuild take sufficiently long for the tests. The quickest I have seen a volume this size being rebuilt is approximately 30 seconds.